### ISSUE TITLE
How to set up database for contributing?

### ISSUE THREAD

**Issue by ab-pm** [graphile/postgraphile#1136]
Hi,
I've read [the contributing document](https://github.com/graphile/postgraphile/blob/master/CONTRIBUTING.md) and think I've set up the js development tools fine, but I can't get the tests to run.

It might have been obvious, but should still be explicitly documented that one needs to have Postgres installed :-) Then, in addition to the repository cloning and setup lines, I did run
```bash
sudo apt update
sudo apt-get install postgresql postgresql-contrib postgresql-client-common
sudo -u postgres createdb postgraphile_test
sudo -u postgres createuser mytestuser
sudo -u postgres psql
	GRANT ALL PRIVILEDGES ON DATABASE postgraphile_test TO mytestuser;
	GRANT pg_read_all_settings TO mytestuser;
	ALTER ROLE mytestuser WITH CREATEDB;
	ALTER ROLE mytestuser WITH REPLICATION;
	\q
npm config set scripts-prepend-node-path auto
sudo nano /etc/postgresql/11/main/postgresql.conf
	set wal_level to 'logical' instead of 'replica'
export TEST_DATABASE_URL="postgres://localhost/postgraphile_test"
```
When I just ran `yarn test` in the `/graphile-engine` directory, I hadn't noticed that there are [separate setup instructions for the `lds` package](https://github.com/graphile/graphile-engine/tree/master/packages/lds), so I was greeted with errors first until I added the required privileges.

I'm stuck now with the other tests, though. The `lds` test reports "*error: password authentication failed for user "mytestuser"*" (from `pg/lib/connection.js`). If I "[focus on the integration tests](https://github.com/graphile/postgraphile/blob/master/CONTRIBUTING.md#graphile-engine-2)" and run `yarn test queries` in `packages/postgraphile-core`, I am greeted with a password prompt for mytestuser and when I set one up and enter it there, I get
```
Password for user mytestuser: 
NOTICE:  schema "a" does not exist, skipping
NOTICE:  schema "b" does not exist, skipping
NOTICE:  schema "c" does not exist, skipping
NOTICE:  schema "d" does not exist, skipping
ERROR:  permission denied to drop role
error Command failed with exit code 3.
```
What other role do I need? And how to run tests without entering my password every time? Do I need to set up a [`.pgpass` file](https://www.postgresql.org/docs/current/libpq-pgpass.html)? Or change the `pg_hba.conf`iguration?

You can assume that `mytestuser` is my Ubuntu 18.04 username, if that helps.

**Comment by benjie** [comment]
I'd really appreciate if you could expand the CONTRIBUTING document with the parts that are missing once you've figured them out.

The tests currently assume that you're running PostgreSQL locally, and that you have 'trust' auth enabled such that your user is automatically used for connection. On Mac this happened for me automatically, but I recently moved to ubuntu and I now have to do this manually. I have the following script I added to `~/.env` which I now source whenever I need it (before running the tests) `. ~/.env`:

```
export PGHOST='/var/run/postgresql'
export TEST_DATABASE_URL="postgres:///pggql_test" 
```

I'm not sure why the PGHOST is required on Linux but not on Mac.

You'll know it's working if you can run `psql template1` and it just works (without asking for credentials). I believe I also created a postgresql superuser that shares my Linux username.

Let me know if this gets you any further, and where you need help next. :pray:

**Comment by ab-pm** [comment]
Ah, needing to run the tests as a (db) superuser is good to know, then I wouldn't have had to grant the rights individually.
I'll follow your approach with the `.env` script, I currently used my `.bashrc` which is probably not the best idea. I'll try this tomorrow and let you know if it doesn't work - or I'll write a pull request with updated documentation :-)

**Comment by benjie** [comment]
Awesome; thanks @ab-pm!

**Comment by akaRem** [comment]
You can just 
`docker run --name test_db_container -p 5432:5432 -h localhost -e POSTGRES_USER=postgres -e POSTGRES_PASSWORD=postgres -e POSTGRES_DB=postgres --rm postgres:10.7-alpine`. 

The connection will be 
```js
new Pool({
  host: "localhost", 
  port: 5432, 
  user: "postgres", 
  password: "'postgres", 
  database: "postgres" 
}
```
(you can easily change values (it's actually container defaults) in docker run)

DB spins up in 2-5 seconds, it's clean and ready for experiments.

**Comment by benjie** [comment]
Yep, running it via Docker should work too, but you have to use a PostgreSQL container with wal2json and logical decoding enabled (not just the “postgres” one). Also if you work on a lot of projects in parallel you either need a global postgres container everyone uses, or you need to use different postgres ports for each app.

**Comment by akaRem** [comment]
I haven't reach a point w/ extended subscriptions where wal2json is required, so it works for me now. I don't think that docker w/ wal2json is something hard to build.
Anyways, my conclusion is that dockerized postgres is very easy to start.

**Comment by ab-pm** [comment]
OK, I didn't got to work on this last week, and yesterday I lost half a day with figuring out that I was connecting to the wrong database - the server I had wanted to use did listen on a different port :-/
Now I hope I found all the environment variables that are responsible for the test database connections, and documented them.

**Comment by ab-pm** [comment]
Found another thing. When running `postgraphile-core` tests, specifically `integration/queries` and `integration/mutations`, I was getting spurious snapshot failures: expecting `"money": 9876543.21,` but receiving `"money": 987654321,`. Turns out it did work when running the tests against a docker-hosted database, but not in my local installation. After some research, it turns out that the [`money` type precision depends on settings](https://www.postgresql.org/docs/current/datatype-money.html), and indeed my local database installation had `lc_monetary = de_DE.UTF-8`, while the docker one had the default `C` (whatever that means). After fixing my `postgresql.conf`, the tests are working now.

However, I don't really feel qualified to write a database setup instruction for this. Presumably some of the [other locale settings](https://www.postgresql.org/docs/current/runtime-config-client.html#RUNTIME-CONFIG-CLIENT-FORMAT) have to be adjusted as well? Which of them could affect tests?

**Comment by benjie** [comment]
That money issue _might_ be a bug. We've code that tweaks money to numeric and then to text:

https://github.com/graphile/graphile-engine/blob/e805289405ad344a1f8329e19864989f53802a20/packages/graphile-build-pg/src/plugins/PgTypesPlugin.js#L253-L254

My guess is that it's actually an input issue in the test fixtures though; triggered here:

https://github.com/graphile/graphile-engine/blob/254c80a85f16fd98c599a536abe86e72369aeebf/packages/postgraphile-core/__tests__/kitchen-sink-data.sql#L89

or here:

https://github.com/graphile/graphile-engine/blob/254c80a85f16fd98c599a536abe86e72369aeebf/packages/postgraphile-core/__tests__/kitchen-sink-data.sql#L124

If you could dig into that further, that'd be great. It might be that we have to cast it to numeric and then have Postgres cast that numeric to money rather than going to money directly from text as it does currently.

**Comment by benjie** [comment]
I'm not sure which other local settings will break the test DB; it's designed to run against default PostgreSQL settings (with UTC timezone and modifications for logical decoding), writing an exhaustive list of the settings that might break the tests and then being sure to keep that up to date would be quite an undertaking.

**Comment by ab-pm** [comment]
Inspired by [`run-pg`](https://github.com/graphile/postgraphile/blob/295e612020c60ccc8bd4eff02d4f4e85a2e3fc8c/run-pg) I now came up with a complete docker setup: https://gist.github.com/ab-pm/2ac739c98d2fb5629cb8bb42ccaa7537

**Comment by benjie** [comment]
It seems a shame to lose this in a gist... Would you like to suggest in a new issue or pull request how this could be added to the `graphile-engine` repo without being too noisy?

**Comment by ab-pm** [comment]
Not sure, maybe put it in the wiki (next to https://github.com/graphile/graphile-engine/wiki/Development-with-docker-compose) and link it from the setup instructions?

**Comment by benjie** [comment]
:+1:
